File: v1/scripts/dedupe.py
"""
Title+author normalization for dedupe.
Mirror of web/lib/dedupe.ts — keep them in sync.
"""

from __future__ import annotations

import re


def _normalize(s: str) -> str:
    s = s.lower().replace("&", "and")
    s = re.sub(r"[^\w\s]", " ", s)
    s = re.sub(r"\s+", " ", s).strip()
    return s


def _normalize_author(raw: str) -> str:
    s = raw.strip()
    if "," in s and s.count(",") == 1:
        last, first = (p.strip() for p in s.split(","))
        # Treat "Last, First" as a single inverted author, but avoid
        # rewriting comma-separated coauthors like "Marily Nika, Diego Granados".
        if last and first and not (len(last.split()) > 1 and len(first.split()) > 1):
            s = f"{first} {last}"
    return _normalize(s)


def make_dedupe_key(title: str, author: str) -> str:
    return f"{_normalize(title)}::{_normalize_author(author)}"

File: v1/web/lib/dedupe.ts
// Normalize title + author into a stable dedupe key.
// The same logic is mirrored in scripts/ingest.py — keep them in sync.
//
// Handles common EPUB metadata messiness:
// - "Last, First" → "First Last"
// - case differences
// - extra whitespace
// - surrounding punctuation/quotes
// - "&" vs "and"
//
// This is best-effort, not perfect. If it fails to dedupe, the user
// deletes the duplicate and re-adds. Acceptable for v1.

function normalizeAuthor(raw: string): string {
  let s = raw.trim();
  // "Last, First" → "First Last"
  if (s.includes(",") && s.split(",").length === 2) {
    const [last, first] = s.split(",").map((p) => p.trim());
    // Avoid rewriting comma-separated coauthors like
    // "Marily Nika, Diego Granados".
    if (last && first && !(last.split(/\s+/).length > 1 && first.split(/\s+/).length > 1)) {
      s = `${first} ${last}`;
    }
  }
  return normalize(s);
}

function normalize(s: string): string {
  return s
    .toLowerCase()
    .replace(/&/g, "and")
    .replace(/[^\w\s]/g, " ")  // strip punctuation
    .replace(/\s+/g, " ")
    .trim();
}

export function makeDedupeKey(title: string, author: string): string {
  return `${normalize(title)}::${normalizeAuthor(author)}`;
}
